test-grapesy/Test/Sanity/Exception.hs

```diff
@@ -0,0 +1,189 @@
+          -- New calls still succeed
+          assertBool "" . predicate
+            =<< Client.withRPC conn def (Proxy @Trivial) (countUntil predicate)
+      , server = [
+              Server.someRpcHandler $
+                Server.mkRpcHandler @Trivial $ \call -> do
+                    handlerCount <-
+                      atomicModifyIORef' handlerCounter (\n -> (n + 1, n))
+                    when (handlerCount == 25) $
+                      throwIO $
+                        DeliberateException $ SomeServerException 1
+                    incUntilFinal call
+          ]
+      }
+
+-- | This is essentially 'Test.Prop.Dialogue.earlyTermination15', but the server
+-- does not wait for client termination.
+test_earlyTerminationNoWait :: IO ()
+test_earlyTerminationNoWait = testClientServer $ ClientServerTest {
+      config = def
+    , client = simpleTestClient $ \conn -> do
+        _mResult <-
+          try @DeliberateException $
+            Client.withRPC conn def (Proxy @Trivial) $ \_call ->
+              throwIO (DeliberateException $ SomeServerException 0)
+
+        Client.withRPC conn def (Proxy @Trivial) $ \call -> do
+          Binary.sendFinalInput @Word8 call 0
+          _output <- Binary.recvOutput @Word8 call
+          return ()
+    , server = [
+          Server.someRpcHandler $
+            Server.mkRpcHandler @Trivial $ \call ->
+              Binary.recvInput @Word8 call >>= \case
+                _ -> Server.sendTrailers call NoMetadata
+        ]
+    }
+
+{-------------------------------------------------------------------------------
+  Client and handler functions
+-------------------------------------------------------------------------------}
+
+-- | Send numbers to the server until it responds with one that satisfies the
+-- given predicate.
+countUntil :: (Word64 -> Bool) -> Client.Call Trivial -> IO Word64
+countUntil = go 0
+  where
+    go :: Word64 -> (Word64 -> Bool) -> Client.Call Trivial -> IO Word64
+    go next p call
+      | p next
+      = do
+        Binary.sendFinalInput @Word64 call next
+        (_final, NoMetadata) <- Binary.recvFinalOutput @Word64 call
+        return next
+      | otherwise
+      = do
+        Binary.sendNextInput @Word64 call next
+        next' <- Binary.recvNextOutput @Word64 call
+        go next' p call
+
+-- | Reads numbers from the client and sends them back incremented by one.
+incUntilFinal :: Server.Call Trivial -> IO ()
+incUntilFinal call = do
+    Binary.recvInput call >>= \case
+      StreamElem n -> do
+        Binary.sendNextOutput @Word64 call $ succ n
+        incUntilFinal call
+      FinalElem n _ -> do
+        Binary.sendFinalOutput @Word64 call (succ n, NoMetadata)
+      NoMoreElems _ -> do
+        -- TODO: <https://github.com/well-typed/grapesy/issues/209>
+        --
+        -- We shouldn't need to handle this case, since our client never
+        -- explicitly sends 'NoMoreElems'. However, see discussion in the
+        -- ticket above.
```

Comment:
Once the `http2` stuff is merged, this ticket can be closed right?

Reply:
We can do that independently from this PR, of course. Remind me, do we rely on any other recent `http2`/`http-semantics` changes for this PR to go through?